src/database-controller/src/core/k8s.js

```diff
@@ -0,0 +1,250 @@
+  But during the reconnection, listFn will be called again, which is inefficient.
+  According to https://github.com/kubernetes-client/javascript/blob/932c2fbc34db954c6ed397b3cd9ead08b2ff1d10/src/cache.ts#L82-L85,
+  this behavior will be fixed in the future.
+  TO DO: If @kubernetes/client-node fixes this issue, we should upgrade our code to use the new code.
+
+  If the informer encounters any error, it will stop watching, and won't re-connect.
+  One can restart it by informer.on('error', (err) => { informer.start(); }).
+
+  */
+  const listFn = () => {
+    logger.info('Frameworks are listed.');
+    return customObjectsClient.listNamespacedCustomObject(
+      'frameworkcontroller.microsoft.com',
+      'v1',
+      namespace,
+      'frameworks',
+    );
+  };
+  const informer = k8s.makeInformer(
+    kc,
+    `/apis/frameworkcontroller.microsoft.com/v1/frameworks?timeoutSeconds=${timeoutSeconds}`,
+    listFn,
+  );
+  return informer;
+}
+
+const priorityClassClient = kc.makeApiClient(k8s.SchedulingV1Api);
+
+async function createPriorityClass(priorityClassDef) {
+  const res = await priorityClassClient.createPriorityClass(priorityClassDef);
+  return res.response;
+}
+
+async function deletePriorityClass(name) {
+  const res = await priorityClassClient.deletePriorityClass(name);
+  return res.response;
+}
+
+const coreV1Client = kc.makeApiClient(k8s.CoreV1Api);
+
+async function createSecret(secretDef) {
+  const res = await coreV1Client.createNamespacedSecret(
+    secretDef.metadata.namespace,
+    secretDef,
+  );
+  return res;
+}
+
+async function deleteSecret(name, namespace = 'default') {
+  const res = await coreV1Client.deleteNamespacedSecret(name, namespace);
+  return res.response;
+}
+
+async function patchSecretOwnerToFramework(secret, frameworkResponse) {
+  const metadata = {
+    ownerReferences: [
+      {
+        apiVersion: 'frameworkcontroller.microsoft.com',
+        kind: 'Framework',
+        name: frameworkResponse.metadata.name,
+        uid: frameworkResponse.metadata.uid,
+        controller: true,
+        blockOwnerDeletion: true,
+      },
+    ],
+  };
+  const res = await coreV1Client.patchNamespacedSecret(
+    secret.metadata.name,
+    secret.metadata.namespace,
+    { metadata: metadata },
+    ...Array(4), // skip some parameters
+    { headers: { 'Content-Type': 'application/merge-patch+json' } },
+  );
+  return res.response;
+}
+
+const timeoutMs = config.k8sConnectionTimeoutSecond * 1000;
+
+// We give every method a timeout.
+module.exports = {
+  getFramework: timeoutDecorator(
+    getFramework,
+    'Kubernetes getFramework',
+    timeoutMs,
+  ),
+  listFramework: timeoutDecorator(
+    listFramework,
+    'Kubernetes getFramework',
+    timeoutMs,
+  ),
+  createFramework: timeoutDecorator(
+    createFramework,
+    'Kubernetes createFramework',
+    timeoutMs,
+  ),
+  patchFramework: timeoutDecorator(
+    patchFramework,
+    'Kubernetes patchFramework',
+    timeoutMs,
+  ),
```

Comment:
Does the k8s client already support timeout? If so why need implement timeoutDecorator

Reply:
I didn't find a global support for timeout in https://github.com/kubernetes-client/javascript